pilot/pkg/networking/core/v1alpha3/listener.go

```diff
@@ -2078,16 +2240,66 @@ func buildCompleteFilterChain(pluginParams *plugin.InputParams, mutable *plugin.
 	}
 
 	httpConnectionManagers := make([]*http_conn.HttpConnectionManager, len(mutable.FilterChains))
+	thriftProxies := make([]*thrift_proxy.ThriftProxy, len(mutable.FilterChains))
 	for i := range mutable.FilterChains {
 		chain := mutable.FilterChains[i]
 		opt := opts.filterChainOpts[i]
 		mutable.Listener.FilterChains[i].Metadata = opt.metadata
 
-		// we are building a network filter chain (no http connection manager) for this filter chain
-		// In HTTP, we need to have mixer, RBAC, etc. upfront so that they can enforce policies immediately
-		// For network filters such as mysql, mongo, etc., we need the filter codec upfront. Data from this
-		// codec is used by RBAC or mixer later.
-		if opt.httpOpts == nil {
+		if opt.thriftOpts != nil && features.EnableThriftFilter.Get() {
+			var quotas []model.Config
+			// Add the TCP filters first.. and then the Thrift filter
+			mutable.Listener.FilterChains[i].Filters = append(mutable.Listener.FilterChains[i].Filters, chain.TCP...)
+
+			thriftProxies[i] = buildThriftProxy(opt.thriftOpts)
+
+			if pluginParams.Service != nil {
+				quotas = opts.push.IstioConfigStore.QuotaSpecByDestination(&model.ServiceInstance{
+					Service: pluginParams.Service,
+				})
+			}
+
+			// If the RLS service was provided, add the RLS to the Thrift filter
+			// chain. Rate limiting is only applied client-side.
+			if rlsURI := features.ThriftRatelimitService.Get(); rlsURI != "" &&
+				mutable.Listener.TrafficDirection == core.TrafficDirection_OUTBOUND &&
+				pluginParams.Service != nil &&
+				pluginParams.Service.Hostname != "" &&
+				len(quotas) > 0 {
+				rateLimitConfig := buildThriftRatelimit(rlsURI, fmt.Sprint(pluginParams.Service.Hostname))
+				rateLimitFilter := &thrift_proxy.ThriftFilter{
+					Name: "envoy.filters.thrift.rate_limit",
```

Comment:
Seems like something we may want to expose by the extensions API @kyessenov @mandarjog